contracts/beacon/Voucher.sol

```diff
@@ -233,12 +306,60 @@ contract Voucher is OwnableUpgradeable, IVoucher {
         $._authorizedAccounts[account] = isAuthorized;
     }
 
+    /**
+     * Ask VoucherHub to refund voucher for a failed task and
+     * send non-sponsored part back to the requester when needed.
+     * @param voucherHub hub
+     * @param iexecPoco address of PoCo contract
+     * @param taskId id of the task
+     * @param dealId task's deal id
+     * @param dealVolume number of tasks in the deal
+     * @param dealPrice price paid at match orders
+     * @param requester of the task
+     */
+    function _refundVoucherAndRequester(
+        IVoucherHub voucherHub,
+        address iexecPoco,
+        bytes32 taskId,
+        bytes32 dealId,
+        uint256 dealVolume,
+        uint256 dealPrice,
+        address requester
+    ) private {
+        VoucherStorage storage $ = _getVoucherStorage();
+        require(!$._refundedTasks[taskId], "Voucher: task already claimed");
+        $._refundedTasks[taskId] = true;
+        if (dealPrice != 0) {
+            uint256 dealSponsoredAmount = $._sponsoredAmounts[dealId];
+            // The division yields no remainder since dealPrice = taskPrice * volume.
+            uint256 taskPrice = dealPrice / dealVolume;
+            // A positive remainder is possible when the voucher balance is less than
+            // the sponsorable amount. Min(balance, dealSponsoredAmount) is computed
+            // at match orders.
+            // TODO !! do something with the remainder.
+            uint256 taskSponsoredAmount = dealSponsoredAmount / dealVolume;
```

Comment:
Something like:
```suggestion
            uint256 taskSponsoredAmount = dealSponsoredAmount / dealVolume;
            // If some, refund crumbs within the latest claimed task of the Bag-of-Task
            uint256 taskSponsoredAmountWithCrumbs = taskSponsoredAmount 
                + (dealPrice % dealVolume);
            if (taskSponsoredAmountWithCrumbs == dealSponsoredAmount) {
                taskSponsoredAmount = taskSponsoredAmountWithCrumbs;
            }
```

Reply:
I don't think this works because we don't know if other tasks will fail.
IMO let's think this in a separate PR.
Thank you anyway for the comment, it made me find a bug.